src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -23,26 +32,37 @@ class SourceRootStrippedSources:
 class StripSourceRootsRequest:
   """A request to strip source roots for every file in the snapshot.
 
-  The field `representative_path` is used to determine the source root for the files to be stripped.
-  This assumes that every file shares the same source root, which should be true in practice as the
-  `sources` field for a target always has files in the same source root. We don't proactively
-  validate this assumption because of the performance implications of running
-  `SourceRoots.find_by_path` on every single file in the snapshot, as opposed to only one file. See
-  https://github.com/pantsbuild/pants/pull/9112#discussion_r377999025 for more context on this
-  design.
+  The call site must decide whether to allow the snapshot to include multiple source roots or not,
+  e.g. the snapshot including `src/python/lib.py` and `src/java/lib.java`. If the snapshot
+  is certain to only one have single source root among all the files, then the call site should pass
+  a `representative_path` for better performance. If the snapshot is likely to have more than one
+  snapshot, the call site should set `support_multiple_source_roots=True`, which has worse
```

Comment:
Hm one stale reference to `support_multiple_source_roots`, which is now simply `multiple_source_roots`. I'll update this before merging but am going to wait for now to avoid a CI cycle.